### PR TITLE
Mark exceptions with CopyableThrowable to help stacktrace recovery

### PR DESCRIPTION
**Pull request by e5l**

### CODE REVIEW

File: ktor-client/ktor-client-okhttp/jvm/src/io/ktor/client/engine/okhttp/OkHttpWebsocketSession.kt [hunk 1]
```diff
@@ -111,4 +111,10 @@ internal class OkHttpWebsocketSession(
 }
 
 @Suppress("KDocMissingDocumentation")
-class UnsupportedFrameTypeException(frame: Frame) : IllegalArgumentException("Unsupported frame type: $frame")
+class UnsupportedFrameTypeException(
+    private val frame: Frame
```
Comment (cy6erGn0m) [inline ktor-client/ktor-client-okhttp/jvm/src/io/ktor/client/engine/okhttp/OkHttpWebsocketSession.kt]: Not sure how it may work in future _mt_ coroutines

**Comment by cy6erGn0m** [comment]
So, generally saying, having custom exception properties forces us to implement `CopyableThrowable`.

File: ktor-client/ktor-client-okhttp/jvm/src/io/ktor/client/engine/okhttp/OkHttpWebsocketSession.kt [hunk 2]
```diff
@@ -111,4 +111,10 @@ internal class OkHttpWebsocketSession(
...
+) : IllegalArgumentException("Unsupported frame type: $frame"), CopyableThrowable<UnsupportedFrameTypeException> {
+    override fun createCopy(): UnsupportedFrameTypeException? = UnsupportedFrameTypeException(frame).also {
+        it.initCause(this)
```
Comment (dmitrievanthony) [inline ktor-client/ktor-client-okhttp/jvm/src/io/ktor/client/engine/okhttp/OkHttpWebsocketSession.kt]: Just to understand, do we actually need to call `it.initCause(this)`, not `it.initCause(this.cause)`?

File: ktor-server/ktor-server-core/jvm/src/io/ktor/features/CallId.kt [hunk 3]
```diff
@@ -29,7 +30,11 @@ typealias CallIdVerifier = (String) -> Boolean
  * An exception that could be thrown to reject a call due to illegal call id
  * @param illegalCallId that caused rejection
  */
-class RejectedCallIdException(val illegalCallId: String) : IllegalArgumentException()
+class RejectedCallIdException(
+    val illegalCallId: String
+) : IllegalArgumentException(), CopyableThrowable<RejectedCallIdException> {
+    override fun createCopy(): RejectedCallIdException? = RejectedCallIdException(illegalCallId)
```
Comment (dmitrievanthony) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/features/CallId.kt]: Did we forget to add `initCause` here?

File: ktor-server/ktor-server-host-common/jvm/src/io/ktor/server/engine/BaseApplicationResponse.kt [hunk 4]
```diff
@@ -239,21 +242,39 @@ abstract class BaseApplicationResponse(override val call: ApplicationCall) : App
      * [OutgoingContent] is trying to set some header that is not allowed for this content type.
      * For example, only upgrade content can set `Upgrade` header.
      */
-    class InvalidHeaderForContent(name: String, content: String) : IllegalStateException("Header $name is not allowed for $content")
+    class InvalidHeaderForContent(
+        private val name: String, private val content: String
```
Comment (dmitrievanthony) [inline ktor-server/ktor-server-host-common/jvm/src/io/ktor/server/engine/BaseApplicationResponse.kt]: Looks like this class is public, does this change break compatibility (I mean `public` -> `private` replacement)?

Comment (e5l) [reply ktor-server/ktor-server-host-common/jvm/src/io/ktor/server/engine/BaseApplicationResponse.kt]: There were no extra fields in this class before, only constructor params

File: ktor-client/ktor-client-okhttp/jvm/src/io/ktor/client/engine/okhttp/OkHttpWebsocketSession.kt (hunk 2 above)
Comment (e5l) [reply ktor-client/ktor-client-okhttp/jvm/src/io/ktor/client/engine/okhttp/OkHttpWebsocketSession.kt]: Yep, we should do it as the reflective approach in `kotlinx.coroutines`.

**Pull request merged**